paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/source/reader/CDCRecordsWithSplitIds.java

```diff
@@ -0,0 +1,140 @@
+ */
+public class CDCRecordsWithSplitIds implements RecordsWithSplitIds<RecordIterator<Event>> {
+
+    @Nullable private String splitId;
+
+    @Nullable private Reference<RecordIterator<Event>> recordsForSplitCurrent;
+
+    @Nullable private final RecordIterator<Event> recordsForSplit;
+
+    private final Set<String> finishedSplits;
+
+    private CDCRecordsWithSplitIds(
+            @Nullable String splitId,
+            @Nullable RecordIterator<Event> recordsForSplit,
+            Set<String> finishedSplits) {
+        this.splitId = splitId;
+        this.recordsForSplit = recordsForSplit;
+        this.finishedSplits = finishedSplits;
+    }
+
+    @Nullable
+    @Override
+    public String nextSplit() {
+        // move the split one (from current value to null)
+        final String nextSplit = this.splitId;
+        this.splitId = null;
+
+        // move the iterator, from null to value (if first move) or to null (if second move)
+        this.recordsForSplitCurrent =
+                nextSplit != null ? new Reference<>(this.recordsForSplit) : null;
+
+        return nextSplit;
+    }
+
+    @Nullable
+    @Override
+    public RecordIterator<Event> nextRecordFromSplit() {
+        if (this.recordsForSplitCurrent == null) {
+            throw new IllegalStateException();
+        }
+
+        RecordIterator<Event> recordsForSplit = this.recordsForSplitCurrent.get();
+        this.recordsForSplitCurrent.set(null);
+        return recordsForSplit;
+    }
+
+    @Override
+    public Set<String> finishedSplits() {
+        return finishedSplits;
+    }
+
+    @Override
+    public void recycle() {
+        if (recordsForSplit != null) {
+            recordsForSplit.releaseBatch();
+        }
+    }
+
+    public static CDCRecordsWithSplitIds forRecords(
+            String splitId, RecordIterator<Event> recordsForSplit) {
+        return new CDCRecordsWithSplitIds(splitId, recordsForSplit, Collections.emptySet());
+    }
+
+    public static CDCRecordsWithSplitIds finishedSplit(String splitId) {
+        return new CDCRecordsWithSplitIds(null, null, Collections.singleton(splitId));
+    }
+
+    public static void emitRecord(
+            SourceReaderContext context,
+            RecordIterator<Event> element,
+            SourceOutput<Event> output,
+            FileStoreSourceSplitState state,
+            FileStoreSourceReaderMetrics metrics) {
+        long timestamp = TimestampAssigner.NO_TIMESTAMP;
+        if (metrics.getLatestFileCreationTime() != FileStoreSourceReaderMetrics.UNDEFINED) {
+            timestamp = metrics.getLatestFileCreationTime();
+        }
+
+        // This metric only counts the number of RecordIterator<Event> emitted,
+        // however what we really want is to count the number of Events emitted,
+        // so we replenish the missing record count here.
+        org.apache.flink.metrics.Counter numRecordsIn =
+                context.metricGroup().getIOMetricGroup().getNumRecordsInCounter();
```

Comment:
The Counter here should be at the Source level. 
Should we provide metrics for individual tables?

Reply:
Same as above, this PR has been big enough for reviewers and I prefer to have detailed metrics supported in future PRs.